esx_service/vmdk_ops.py

```diff
@@ -777,31 +776,90 @@ def disk_attach(vmdk_path, vm):
         except vim.fault.VimFault as ex:
             msg=("Failed to add PVSCSI Controller: %s", ex.msg)
             return err(msg)
+            
         # Find the controller just added
         devices = vm.config.hardware.device
         pvsci = [d for d in devices
                  if type(d) == vim.ParaVirtualSCSIController and
                  d.key == controller_key]
-        pci_slot_number = get_controller_pci_slot(vm, pvsci,
+        pci_slot_number = get_controller_pci_slot(vm, pvsci[0],
                                                   offset_from_bus_number)
 
     # Find a slot on the controller, issue attach task and wait for completion
     if not disk_slot:
-        taken = set([dev.unitNumber
+        idx = 0
+           
+        while ((disk_slot is None) and (idx < len(pvsci))):
+            controller_key = pvsci[idx].key
+            taken = set([dev.unitNumber
                      for dev in devices
                      if type(dev) == vim.VirtualDisk and dev.controllerKey ==
                      controller_key])
-        # search in 15 slots, with unit_number 7 reserved for scsi controller
-        availSlots = (set(range(0, 7)) | set(range(8, 16))) - taken
-
-        if len(availSlots) == 0:
-            msg = "Failed to place new disk - out of disk slots"
-            logging.error(msg + " VM=%s", vm.config.uuid)
-            return err(msg)
-
-        disk_slot = availSlots.pop()
-        logging.debug("controller_key = %d slot = %d", controller_key, disk_slot)
+            # search in 15 slots, with unit_number 7 reserved for scsi controller
+            availSlots = (set(range(0, 7)) | set(range(8, 16))) - taken
+            logging.debug("idx=%d controller_key=%d availSlots=%d", idx, controller_key, len(availSlots))
+
+            if len(availSlots) == 0:
+                idx = idx+1
+            else:     
+                disk_slot = availSlots.pop()
+                pci_slot_number = get_controller_pci_slot(vm, pvsci[idx],
+                                                          offset_from_bus_number)
+                 
+                logging.debug("Find an available slot: controller_key = %d slot = %d", controller_key, disk_slot)
+
+        # search through all the existing pvscsi controller, and cannot find a slot
+        # try to add another pvscsi controller if possible        
+        if ((disk_slot is None) and (len(availSlots) == 0)):
```

Comment:
Since this code is being changed please make the change to rename availSlots to the naming style of "avail_slots" or "available_slots"